### PR TITLE
perf(android-views): Avoid inflating views where possible when binding calendar pages

### PR DESCRIPTION
**Pull request by boswelja**
View inflation is expensive, we should try to avoid that when binding

### CODE REVIEW

**Comment by boswelja** [comment]
As of now the average CPU usage seems to have dropped from 14% to 10%
Device: Pixel 6
Android Version: 12.1 Beta
Note this was measured in the debug variant of the sample app.

**Comment by boswelja** [comment]
Memory graph is looking significantly better (much less allocations on scroll), and CPU usage is significantly lower (around 8% now). A change somewhere seems to have broken height increases caused by page source changes. Investigating now.

**Comment by boswelja** [comment]
CPU utilisation increased when page source changes. This is because we're now invalidating (most of) the cache when that happens. Date cells are still retained, if possible

File: android/views/src/main/kotlin/com/boswelja/ephemeris/views/CalendarPagerAdapter.kt [hunk 1]
```diff
@@ -49,50 +48,106 @@ internal class CalendarPagerAdapter : InfinitePagerAdapter<CalendarPageViewHolde
 
     override fun onBindHolder(holder: CalendarPageViewHolder, page: Int) {
         val pageState = pageLoader!!.getPageData(page)
-        holder.bindDisplayRows(dayBinder!!, pageState)
+        holder.bindDisplayRows(pageLoader!!, dayBinder!!, pageState)
```
Comment (Iannnr) [inline android/views/src/main/kotlin/com/boswelja/ephemeris/views/CalendarPagerAdapter.kt:51]: is this still needed after `pageLoader!!.getPageData(page)`? Is it not smart casting?

File: android/views/src/main/kotlin/com/boswelja/ephemeris/views/CalendarPagerAdapter.kt [hunk 2]
```diff
@@ -49,50 +48,106 @@ internal class CalendarPagerAdapter : InfinitePagerAdapter<CalendarPageViewHolde
...
     private val binding: CalendarPageBinding
-) : RecyclerView.ViewHolder(binding.root) {
+) : ViewHolder(binding.root) {
 
     private val inflater = LayoutInflater.from(itemView.context)
 
+    private val rowBindingCache = mutableListOf<CalendarRowBinding>()
+    private val dateCellViewHolderCache = mutableListOf<ViewHolder>()
+
+    private var boundDateCells = 0
```
Comment (Iannnr) [inline android/views/src/main/kotlin/com/boswelja/ephemeris/views/CalendarPagerAdapter.kt:64]: is this actually any different to
```kotlin
private val boundDateCells: Int
    get() = dateCellViewHolderCache.size
````
Seems a bit excessive to keep track of 2 variables for this? Or is it purely for calculating how many to drop when you're trimming?

File: android/views/src/main/kotlin/com/boswelja/ephemeris/views/CalendarPagerAdapter.kt [hunk 3]
```diff
@@ -49,50 +48,106 @@ internal class CalendarPagerAdapter : InfinitePagerAdapter<CalendarPageViewHolde
...
+                // Clear our cached date cells on binder change
+                dateCellViewHolderCache.clear()
+            }
+        }
+    private var pageLoader: CalendarPageLoader? = null
+        set(value) {
+            if (field != value) {
+                field = value
+                // Only invalidate row cache. date cells are trimmed after binding to optimize CPU
+                rowBindingCache.removeAll {
```
Comment (Iannnr) [inline android/views/src/main/kotlin/com/boswelja/ephemeris/views/CalendarPagerAdapter.kt:80]: I'm tempted to say 
```kotlin
rowBindingCache.onEach { 
    it.root.removeAllViewsInLayout() 
}.clear()
```
Will be better here. We don't need the predicate to decide if we're always clearing the whole list, so there's no extra decisions to make on each iteration

File: android/views/src/main/kotlin/com/boswelja/ephemeris/views/pager/InfiniteAnimatingPager.kt [hunk 4]
```diff
@@ -91,13 +93,43 @@ internal class PageChangeAnimator(
...
-                it.height = animator.animatedValue as Int
+        if (fromHeight != toHeight) {
+            heightAnimator.setIntValues(fromHeight, toHeight)
+            heightAnimator.addUpdateListener { animator ->
+                newHolder.itemView.layoutParams = newHolder.itemView.layoutParams.also {
+                    it.height = animator.animatedValue as Int
+                }
             }
+            heightAnimator.addListener(
+                object : AnimatorListener {
```
Comment (Iannnr) [inline android/views/src/main/kotlin/com/boswelja/ephemeris/views/pager/InfiniteAnimatingPager.kt:104]: I wonder if there exists an AnimatorListener which acts like the core-ktx 
```
public inline fun TextView.doOnTextChanged(
    crossinline action: (
        text: CharSequence?,
        start: Int,
        before: Int,
        count: Int
    ) -> Unit
): TextWatcher = addTextChangedListener(onTextChanged = action)
```
So you only need to implement doOnAnimationEnd and doOnAnimationCancel so we can omit the no-ops 🤔

File: android/views/src/main/kotlin/com/boswelja/ephemeris/views/CalendarPagerAdapter.kt (hunk 2 above)
Comment (boswelja) [reply android/views/src/main/kotlin/com/boswelja/ephemeris/views/CalendarPagerAdapter.kt:64]: It's there to track which date holder to try look up from the cache, and it's reused when trimming yep

File: android/views/src/main/kotlin/com/boswelja/ephemeris/views/CalendarPagerAdapter.kt (hunk 1 above)
Comment (boswelja) [reply android/views/src/main/kotlin/com/boswelja/ephemeris/views/CalendarPagerAdapter.kt:51]: pageLoader is mutable, therefore could change at any time (effectively impossible in this case though)

File: android/views/src/main/kotlin/com/boswelja/ephemeris/views/CalendarPagerAdapter.kt (hunk 3 above)
Comment (boswelja) [reply android/views/src/main/kotlin/com/boswelja/ephemeris/views/CalendarPagerAdapter.kt:80]: I figured `clear` iterates the whole list again, but if that's not the case then `onEach` absolutely makes more sense

File: android/views/src/main/kotlin/com/boswelja/ephemeris/views/pager/InfiniteAnimatingPager.kt (hunk 4 above)
Comment (boswelja) [reply android/views/src/main/kotlin/com/boswelja/ephemeris/views/pager/InfiniteAnimatingPager.kt:104]: There is not sadly, we'd need to include core-ktx for that. All for cleaning it up if you've got any ideas, it's kinda on the larger side

File: android/views/src/main/kotlin/com/boswelja/ephemeris/views/CalendarPagerAdapter.kt (hunk 3 above)
Comment (Iannnr) [reply android/views/src/main/kotlin/com/boswelja/ephemeris/views/CalendarPagerAdapter.kt:80]: ah yes it does, 
```java
// clear to let GC do its work
for (int i = 0; i < size; i++)
    elementData[i] = null;

size = 0;
```
love it, just nulls out each element, sets the list size to 0 and lets GC clear it :joy:

Comment (boswelja) [reply android/views/src/main/kotlin/com/boswelja/ephemeris/views/CalendarPagerAdapter.kt:80]: Yep in that case iterating once should be quicker, would be nice if we didn't need the predicate but I don't see an existing extension for it sadly

File: android/views/src/main/kotlin/com/boswelja/ephemeris/views/pager/InfiniteAnimatingPager.kt (hunk 4 above)
Comment (Iannnr) [reply android/views/src/main/kotlin/com/boswelja/ephemeris/views/pager/InfiniteAnimatingPager.kt:104]: https://developer.android.com/kotlin/ktx/extensions-list#for_androidanimationanimator damn there already is a ktx extension for it, they beat me to it :( my only suggestion to clean it up would be that

**Review by Iannnr — APPROVED** [review]

File: android/views/src/main/kotlin/com/boswelja/ephemeris/views/pager/InfiniteAnimatingPager.kt (hunk 4 above)
Comment (boswelja) [reply android/views/src/main/kotlin/com/boswelja/ephemeris/views/pager/InfiniteAnimatingPager.kt:104]: I'm allergic to adding extra deps, maybe we can create one though

Comment (Iannnr) [reply android/views/src/main/kotlin/com/boswelja/ephemeris/views/pager/InfiniteAnimatingPager.kt:104]: hah, yeah can do that as a special case for us

**Pull request merged**